### ISSUE TITLE
Performing optimizations in parallel

### ISSUE THREAD

**Issue by emauryg** [mlverse/torch#443]
I'm trying to optimize a optimize a tensor in some sort map reduce approach using torch in R.  In the code below is a schematic of the for loop implementation, but ideally i'd be able to run these in batches of S rather than one sample at a time.  Is there a way currently to implement this with torch? A naive approach might be to create some sort of dataset object. 

```
optimize_theta <- function(T0, covs, eta, Sigma, K, S, count_matrix,mu, learning_rate=0.2){
  nu = torch_empty(c(S,K-1, K-1), device=device)
  theta = torch_empty(c(K,S), device=device)
  T_tmp = T0toT(T0)
  f = covs_to_F(covs$bt, covs$br, covs$epi, covs$nuc, covs$clu)
  T_tensor = T_tmp$mul(f$unsqueeze(-2))
  for (s in 1:S){
    ## parallelize code by samples to reduce estimation variance
    old_loss = 1e11
    eta_s = eta[,s]
    mu_s = mu[,s]
    Y_d = count_matrix[,,,,,,s]
    MODEL = estimate_theta(eta_s, Sigma, mu_s, K, S, Y_d, T_tensor)
    optimizer = optim_adam(MODEL$parameters, lr = learning_rate)
    it=0
    not_converged =TRUE
    while (not_converged == TRUE && it != 1000){
      it = it+1
      theta_est = MODEL()
      optimizer$zero_grad()
      new_loss = theta_est
      new_loss$backward()
      optimizer$step()
      
      # if (it %% 100 == 0 || it==1){
      #   cat("Theta Step:",it,":", as.numeric(new_loss$item()), "\n")
      # }
      
      if(abs(new_loss$item() - old_loss)/old_loss < 1e-2){
        not_converged = FALSE
      } else {
        old_loss = new_loss$item()
      }
    }
    eta[,s] = MODEL$parameters$eta_s$detach()
    nu[s] = calc_hessInv(K, S, eta[,s], T_tensor, Y_d, MODEL$Sigma_inv$detach())
    theta[,s] = nnf_softmax(torch_cat(c(eta[,s], torch_zeros(1, device=device)),dim=1), dim=1)

  }
      
  return(list(eta=eta, nu = nu, theta = theta))
}
```

**Comment by dfalbel** [comment]
If the only objective of parallelizing is reducing estimation variance, you could accumulate a few iterations before running `optimizer$step`. As you can see in the example below, when you do `x$backward()` multiple times without calling `zero_grad()` the gradient is accumulated:

``` r
library(torch)
x <- torch_tensor(1, requires_grad = TRUE)

for (i in 1:5) {
  loss <- 2*x
  loss$backward()
  print(x$grad)
}
#> torch_tensor
#>  2
#> [ CPUFloatType{1} ]
#> torch_tensor
#>  4
#> [ CPUFloatType{1} ]
#> torch_tensor
#>  6
#> [ CPUFloatType{1} ]
#> torch_tensor
#>  8
#> [ CPUFloatType{1} ]
#> torch_tensor
#>  10
#> [ CPUFloatType{1} ]
```

<sup>Created on 2021-01-20 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

If this is for performance, you would probably need to rewrite your code using vectorized operations.

**Comment by emauryg** [comment]
For gradient accumulation do you mean that I should not call zero_grad() until for example 10 loops of the for (s in 1:S) loop or the while loop?

**Comment by dfalbel** [comment]
I would need to understand what's in `estimate_theta()`, why does it depend on `eta_s` at initialization and not on `forward`?
If you could make it depend only on forward, you could make 

```
theta_est = mean(c(MODEL(eta[,1]), MODEL(eta[,2]), ...))
``` 

Not sure if I understood the problem correctly though

**Comment by emauryg** [comment]
eta[,1] etc are the parameters I'm interested on estimating, I guess the mu_s and Y_d's could be on the forward. Does that make more sense?

**Comment by dfalbel** [comment]
In this case I think I would initialize `n` models each with a different value of `s`. Put all parameters the parameters in the optimizer and make `theta_est` be the mean of all of those models. The while loop could then be identical.

**Comment by emauryg** [comment]
oh I think I understand the idea conceptually, but unsure how that could be actually implemented since basically for each value of s I would have to initialize a model and calculate a loss, but also store the paramters somehow? How does optim_adam() know which parameters below to which model? Sorry if this is confusing. Thank you for the help!

**Comment by dfalbel** [comment]
In theory as far as the parameters passed to `optim_*` are unique, it doesn't need to know anything about their location because they are modified in-place, so knowing a reference is enough.

I think it would be easier if you created another `nn_module` that can take multiple columns of `eta[,]` so `model$parameters` would already give you the the complete set of parameters. Inside of this module it could initialize one `MODEL` for each  column as you do in this loop.
